File: app/src/main/java/com/aandssoftware/aandsinventory/database/InventoryDao.java
package com.aandssoftware.aandsinventory.database;

import android.support.annotation.NonNull;

import com.aandssoftware.aandsinventory.listing.ListType;
import com.aandssoftware.aandsinventory.models.CallbackRealmObject;
import com.aandssoftware.aandsinventory.models.InventoryItem;
import com.aandssoftware.aandsinventory.models.InventoryItemHistory;

import io.realm.Case;
import io.realm.Realm;
import io.realm.Realm.Transaction;
import io.realm.Realm.Transaction.OnSuccess;
import io.realm.RealmList;
import io.realm.RealmQuery;
import io.realm.RealmResults;
import io.realm.Sort;

import java.util.HashMap;
import java.util.Iterator;
import java.util.ListIterator;
import java.util.Map.Entry;

public class InventoryDao {
  
  private Realm mRealm;
  
  public InventoryDao(@NonNull Realm realm) {
    mRealm = realm;
  }
  
  
  public RealmResults<InventoryItem> getInventoryItemRecords() {
    return mRealm.where(InventoryItem.class)
        .equalTo("inventoryType", ListType.LIST_TYPE_INVENTORY.ordinal())
        .lessThanOrEqualTo("parentId", 0)
        .findAll();
  }
  
  public RealmResults<InventoryItem> getMaterialRecords() {
    return mRealm.where(InventoryItem.class)
        .equalTo("inventoryType", ListType.LIST_TYPE_MATERIAL.ordinal())
        .lessThanOrEqualTo("parentId", 0)
        .sort("inventoryItemName", Sort.ASCENDING).findAll();
  }
  
  
  public RealmResults<InventoryItem> getInventoryItemRecords(String query) {
    RealmQuery<InventoryItem> realmQuery = mRealm.where(InventoryItem.class).beginGroup()
        .equalTo("inventoryType", ListType.LIST_TYPE_INVENTORY.ordinal())
        .lessThanOrEqualTo("parentId", 0).endGroup();
    return realmQuery.findAll();
  }
  
  public RealmResults<InventoryItem> getInventoryItemRecords(int inventoryType, String query) {
    if (inventoryType == ListType.LIST_TYPE_INVENTORY.ordinal()) {
      if (query != null && !query.isEmpty()) {
        return mRealm.where(InventoryItem.class)
            .equalTo("inventoryType", ListType.LIST_TYPE_INVENTORY.ordinal())
            .lessThanOrEqualTo("parentId", 0).beginGroup()
            .contains("inventoryItemName", query, Case.INSENSITIVE).or()
            .contains("inventoryItemModelName", query, Case.INSENSITIVE).or()
            .contains("inventoryItemBrandName", query, Case.INSENSITIVE).endGroup()
            .sort("inventoryItemName", Sort.ASCENDING).findAll();
      } else {
        return getInventoryItemRecords();
      }
      
    } else {
      if (query != null && !query.isEmpty()) {
        return mRealm.where(InventoryItem.class)
            .equalTo("inventoryType", ListType.LIST_TYPE_MATERIAL.ordinal())
            .lessThanOrEqualTo("parentId", 0).beginGroup()
            .contains("inventoryItemName", query, Case.INSENSITIVE).or()
            .contains("inventoryItemModelName", query, Case.INSENSITIVE).or()
            .contains("inventoryItemBrandName", query, Case.INSENSITIVE).endGroup()
            .sort("inventoryItemName", Sort.ASCENDING).findAll();
      } else {
        return getMaterialRecords();
      }
    }
  }
  
  public RealmResults<InventoryItemHistory> getAllInventoryItemHistory() {
    RealmResults<InventoryItemHistory> realmResultList = mRealm.where(InventoryItemHistory.class)
        .sort("inventoryItemName", Sort.DESCENDING).findAll();
    return realmResultList;
  }
  
  public RealmResults<InventoryItemHistory> getAllInventoryItemHistory(int inventoryId) {
    return mRealm.where(InventoryItemHistory.class).equalTo("id", inventoryId)
        .sort("modifiedDate", Sort.ASCENDING).findAll();
  }
  
  public void saveInventoryItem(InventoryItem mInventoryItem, CallbackRealmObject object) {
    mRealm.executeTransactionAsync(new Transaction() {
      @Override
      public void execute(Realm realm) {
        InventoryItem inventoryItem = new InventoryItem(mInventoryItem);
        realm.copyToRealmOrUpdate(inventoryItem);
      }
    }, new OnSuccess() {
      @Override
      public void onSuccess() {
        object.getCallBack(true);
      }
    });
  }
  
  
  public int getNextInventoryItemId() {
    Number currentIdNum = mRealm.where(InventoryItem.class).max("id");
    int nextId;
    if (currentIdNum == null) {
      nextId = 1;
    } else {
      nextId = currentIdNum.intValue() + 1;
    }
    return nextId;
  }
  
  public int getNextInventoryHistoryId() {
    Number currentIdNum = mRealm.where(InventoryItemHistory.class).max("id");
    int nextId;
    if (currentIdNum == null) {
      nextId = 1;
    } else {
      nextId = currentIdNum.intValue() + 1;
    }
    return nextId;
  }
  
  public void removeInventoryItem(InventoryItem inventoryItem) {
    mRealm.executeTransaction(new Realm.Transaction() {
      @Override
      public void execute(Realm realm) {
        InventoryItem result = mRealm.where(InventoryItem.class)
            .equalTo("id", inventoryItem.getId()).findFirst();
        if (null != result) {
          result.deleteFromRealm();
        }
      }
    });
  }
  
  
  public boolean saveInventoryItemHistory(InventoryItem mInventoryItem,
      HashMap<Integer, InventoryItemHistory> hashMap) {
    try {
      mRealm.executeTransactionAsync(new Transaction() {
        @Override
        public void execute(Realm realm) {
          RealmList<InventoryItemHistory> list = mInventoryItem.getInventoryItemHistories();
          list = (null != list) ? list : new RealmList<>();
          for (Entry<Integer, InventoryItemHistory> integerInventoryItemHistoryEntry : hashMap
              .entrySet()) {
            list.add(integerInventoryItemHistoryEntry.getValue());
          }
          mInventoryItem.setInventoryItemHistories(list);
          realm.copyToRealmOrUpdate(mInventoryItem);
        }
      }, new OnSuccess() {
        @Override
        public void onSuccess() {
        }
      });
      
    } catch (Exception e) {
      e.printStackTrace();
      return false;
    }
    return true;
  }
  
}

File: app/src/main/java/com/aandssoftware/aandsinventory/database/InventoryDao.kt
package com.aandssoftware.aandsinventory.database

import com.aandssoftware.aandsinventory.application.AandSApplication
import com.aandssoftware.aandsinventory.common.Utils
import com.aandssoftware.aandsinventory.firebase.FirebaseUtil
import com.aandssoftware.aandsinventory.firebase.GetAlphaNumericAndNumericIdListener
import com.aandssoftware.aandsinventory.listing.ListType
import com.aandssoftware.aandsinventory.models.CallBackListener
import com.aandssoftware.aandsinventory.models.InventoryItem
import com.aandssoftware.aandsinventory.models.InventoryItemHistory
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.ValueEventListener
import com.google.firebase.storage.FirebaseStorage
import java.util.*


class InventoryDao {

    companion object {
        private var instance: InventoryDao? = null
        @Synchronized
        private fun createInstance() {
            if (instance == null) {
                instance = InventoryDao()
            }
        }

        @JvmStatic
        fun getInstance(): InventoryDao {
            if (instance == null) createInstance()
            return instance as InventoryDao
        }
    }


    constructor() {
    }

    private val inventoryItemTableReference: DatabaseReference
        get() {
            val reference = AandSApplication.getDatabaseInstance()
                    .getReference(InventoryItem.TABLE_INVENTORY_ITEM)
            reference.keepSynced(true)
            return reference.child(Utils.getCustomerIdForCustomerInventory(AandSApplication.instanceApp))
        }

    private val inventoryItemCounterTableReference: DatabaseReference
        get() {
            val reference = AandSApplication.getDatabaseInstance()
                    .getReference(InventoryItem.INVENTORY_ITEM_COUNTER)
            reference.keepSynced(true)
            return reference
        }

    private val materialInventoryItemTableReference: DatabaseReference
        get() {
            val reference = AandSApplication.getDatabaseInstance()
                    .getReference(InventoryItem.TABLE_MATERIAL_INVENTORY_ITEM)
            reference.keepSynced(true)
            return reference
        }

    private val materialInventoryItemCounterTableReference: DatabaseReference
        get() {
            val reference = AandSApplication.getDatabaseInstance()
                    .getReference(InventoryItem.MATERIAL_INVENTORY_ITEM_COUNTER)
            reference.keepSynced(true)
            return reference
        }


    private fun getInventoryItemHistoryTableReference(id: String): DatabaseReference {
        val reference = inventoryItemTableReference.child(id).child(InventoryItem.INVENTORY_ITEM_HISTORY)
        reference.keepSynced(true)
        return reference
    }


    fun getAllInventoryItemAtOnce(lastNodeKey: Double, itemToLoad: Int, valueEventListener: ValueEventListener) {
        var query = inventoryItemTableReference.limitToLast(itemToLoad)
        var defaultLong: Double = 0.0
        query = if (lastNodeKey == defaultLong) {
            query.orderByChild(InventoryItem.ORDER_BY_VALUE).startAt(lastNodeKey)
        } else {
            query.orderByChild(InventoryItem.ORDER_BY_VALUE).endAt(lastNodeKey)
        }
        query.addListenerForSingleValueEvent(valueEventListener)
    }

    fun getAllInventoryItemAtOnce(valueEventListener: ValueEventListener) {
        inventoryItemTableReference.orderByChild(InventoryItem.ORDER_BY_VALUE).addListenerForSingleValueEvent(valueEventListener)
    }


    fun getMaterialRecords(lastNodeKey: Double, itemToLoad: Int, valueEventListener: ValueEventListener) {
        var query = materialInventoryItemTableReference.limitToLast(itemToLoad)
        var defaultLong: Double = 0.0
        query = if (lastNodeKey == defaultLong) {
            query.orderByChild(InventoryItem.ORDER_BY_VALUE).startAt(lastNodeKey)
        } else {
            query.orderByChild(InventoryItem.ORDER_BY_VALUE).endAt(lastNodeKey)
        }
        query.addListenerForSingleValueEvent(valueEventListener)
    }

    fun getNextInventoryItemId(inventoryType: Int,listener: GetAlphaNumericAndNumericIdListener) {
        var counterRef = if (inventoryType == ListType.LIST_TYPE_MATERIAL.ordinal)
            materialInventoryItemCounterTableReference
        else
            inventoryItemCounterTableReference

        var tableRef = if (inventoryType == ListType.LIST_TYPE_MATERIAL.ordinal)
            materialInventoryItemTableReference
        else
            inventoryItemTableReference

        counterRef.addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                var numericId = FirebaseUtil.getInstance().incrementCounter(dataSnapshot)
                var alphaNumericId = tableRef.push().key
                alphaNumericId = alphaNumericId?.let { alphaNumericId } ?: numericId
                listener.afterGettingIds(alphaNumericId, numericId)
            }

            override fun onCancelled(databaseError: DatabaseError) {}
        })
    }


    fun getNextInventoryHistoryId(id: String): String? = getInventoryItemHistoryTableReference(id).push().key


    fun getInventoryItemFromId(inventoryId: String, valueEventListener: ValueEventListener) {
        inventoryItemTableReference.child(inventoryId).addValueEventListener(valueEventListener)
    }

    fun getMaterialInventoryItemFromId(inventoryId: String, valueEventListener: ValueEventListener) {
        materialInventoryItemTableReference.child(inventoryId).addValueEventListener(valueEventListener)
    }


    fun getAllInventoryItemAtOnce(inventoryType: Int, valueEventListener: ValueEventListener) {
        var refrance = if (inventoryType == ListType.LIST_TYPE_MATERIAL.ordinal)
            materialInventoryItemTableReference
        else
            inventoryItemTableReference
        refrance.orderByChild(InventoryItem.ORDER_BY_VALUE_NAME).addListenerForSingleValueEvent(valueEventListener)
    }


    fun getAllInventoryItemHistory(inventoryId: String, valueEventListener: ValueEventListener) {
        getInventoryItemHistoryTableReference(inventoryId).orderByChild(InventoryItemHistory.ORDER_BY_VALUE).addListenerForSingleValueEvent(valueEventListener)
    }

    fun saveInventoryItem(mInventoryItem: InventoryItem, inventoryType: Int, dataListener: CallBackListener) {
        if (null != mInventoryItem.id) {
            var ref = if (inventoryType == ListType.LIST_TYPE_MATERIAL.ordinal)
                materialInventoryItemTableReference
            else
                inventoryItemTableReference

            ref.child(mInventoryItem.id!!).setValue(mInventoryItem) { databaseError, _ ->
                run {
                    if (null == databaseError) {
                        dataListener.getCallBack(true)
                    } else {
                        dataListener.getCallBack(false)
                    }
                }
            }
        } else {
            dataListener.getCallBack(false)
        }
    }

    fun removeInventoryItem(inventoryItem: InventoryItem, inventoryType: Int, completionListener: DatabaseReference.CompletionListener) {
        inventoryItem.id?.let {
            var reference = if (inventoryType == ListType.LIST_TYPE_MATERIAL.ordinal) {
                materialInventoryItemTableReference
            } else {
                inventoryItemTableReference
            }
            reference.child(it).removeValue(completionListener)
        }
    }

    fun removeInventoryImage(url: String, callBackListener: CallBackListener) {
        val photoRef = FirebaseStorage.getInstance().getReferenceFromUrl(url)
        photoRef.delete().addOnSuccessListener { callBackListener.getCallBack(true) }.addOnFailureListener { callBackListener.getCallBack(false) }
    }

    fun saveInventoryItemHistory(alphanumericInventoryId: String, hashMap: HashMap<String, Any>, listner: DatabaseReference.CompletionListener) {
        getInventoryItemHistoryTableReference(alphanumericInventoryId).updateChildren(hashMap, listner)
    }

}
